### DISCUSSION TITLE
The speaker audio of the client is out of sync with the server

### DISCUSSION THREAD

**Discussion by wellcoming** [cc-tweaked/CC-Tweaked#1361] [Q&A]
### Minecraft Version

1.18.x

### Version

1.100.8

### Details

I am making a dual track audio playback program, but due to some factors (It seems that the server is synchronized, but there are some backlogs of audio cache on the client), the two channels will not play at the same time. Is there any way to restrict the two speakers to keep synchronized？

code:
```lua
local dfpwm = require("cc.audio.dfpwm")
local decoder = dfpwm.make_decoder()

f1=fs.open("test1.dfpwm","rb")
f2=fs.open("test2.dfpwm","rb")

local s1_name="speaker_0"
local s2_name="speaker_1"

local s1=peripheral.wrap(s1_name)
local s2=peripheral.wrap(s2_name)

function play1()
    local buffer=decoder(f1.read(1024))
    local name
    while not s1.playAudio(buffer) do
        while name~=s1_name do
            _,name=os.pullEvent("speaker_audio_empty")
        end
        name=nil
    end
end

function play2()
    local buffer=decoder(f2.read(1024))
    local name
    while not s2.playAudio(buffer) do
        while name~=s2_name do
            _,name=os.pullEvent("speaker_audio_empty")
        end
        name=nil
    end
end

while true do
    parallel.waitForAll(play1,play2)
end
```

**Comment by SquidDev** [comment]
> Is there any way to restrict the two speakers to keep synchronized？

So there's no way to _guarantee_ that speakers end up perfectly synchronised. Minecraft's sound engine and the current implementation of speakers mean that sometimes speakers may end up several ticks out of sync.

However, there are definitely ways you can reduce the risk of that happening. In your case, the main problem here is that I don't think `parallel.waitForAll` is doing quite what you expect. `parallel.waitForAll` doesn't run each function at exactly the same time - instead it runs `play1` until it calls `os.pullEvent`, and then switches over to running `play2` (and then back again when that pulls an event). This means your program actually does something like the following:

 1. Reads from `test1.dfpwm` and decodes it.
 2. Plays this decoded audio to `speaker_0`.
 3. _Then_ reads from `test2.dfpwm` and decodes it.
 4. And plays this to `speaker_1`.

As decoding can take a bit of time, there ends up being a bit of a delay between playing the two samples. Instead, you need to decode both samples and then play the audio at the same time:

```lua
-- You actually need a separate decoder for each stream, as otherwise you'll get weird
-- decoding artefacts.
local decoder1 = dfpwm.make_decoder()
local decoder2 = dfpwm.make_decoder()

while true do
  -- I'm also bumping this amount up to 16 * 1024 * 2, which is the maximum you can
  -- play with a speaker in one go - doing bigger batches of audio generally behaves
  -- much better.
  local buffer1 = decoder1(f1.read(16 * 1024 * 2))
  local buffer2 = decoder2(f2.read(16 * 1024 * 2))

  while not s1.playAudio(buffer1) do os.pullEvent("speaker_audio_empty") end
  while not s2.playAudio(buffer2) do os.pullEvent("speaker_audio_empty") end
end
```

**Reply by wellcoming** [reply]
I tested it with your code. First of all, if you use `fs.read` to read the file, the maximum size will be 16*1024  (otherwise it will exceed the buffer limit)  _（Well, I‘m using the fabric version of CC: Tweaked, which may be different from the Forge version）_. Secondly, the audio lag is basically unchanged (about 1~2s), and the audio becomes stuck. Anyway, thank you for your help and hope you can continue helping me.

**Reply by wellcoming** [reply]
After my test, I found that there is no problem at all on Forge 1.19.2. Maybe there are some problems with CC: Restitched. I apologize for my recklessness.

**Reply by SquidDev** [reply]
> The maximum size will be 16*1024

Ahh, thank you! I copied some code from the wrong example in the docs, my mistake - sorry!

> After my test, I found that there is no problem at all on Forge 1.19.2

Oh, that's very odd. 1.18.2 Fabric appears to work for me, so I was wondering if this was a difference further down the audio stack, but if 1.19.2 Forge works for you that's even odder. The two versions should be pretty similar in terms of speaker code, so I'm honestly not sure what's going on.

**Reply by wellcoming** [reply]
This is most likely a bug, I have created a new issue at [here](https://github.com/cc-tweaked/cc-restitched/issues/144)
